jerry-core/rcs/rcs-allocator.cpp

```diff
@@ -0,0 +1,395 @@
+                           record_t *place_p, /**< where to initialize the record */
+                           record_t *next_record_p, /**< next allocated record */
+                           size_t free_size) /**< size of free part between the allocated record
+                                                  and the next allocated record */
+{
+  const size_t node_data_space_size = rcs_get_node_data_space_size ();
+
+  if (next_record_p != NULL)
+  {
+    if (free_size == 0)
+    {
+      rcs_record_set_prev (rec_set_p, next_record_p, place_p);
+    }
+    else
+    {
+      rcs_chunked_list_t::node_t *node_p = rec_set_p->get_node_from_pointer (next_record_p);
+      uint8_t *node_data_space_p = rcs_get_node_data_space (rec_set_p, node_p);
+
+      JERRY_ASSERT ((uint8_t *) next_record_p < node_data_space_p + node_data_space_size);
+
+      record_t *free_rec_p;
+
+      if ((uint8_t *) next_record_p >= node_data_space_p + free_size)
+      {
+        free_rec_p = (record_t *) ((uint8_t *) next_record_p - free_size);
+      }
+      else
+      {
+        size_t size_passed_back = (size_t) ((uint8_t *) next_record_p - node_data_space_p);
+        JERRY_ASSERT (size_passed_back < free_size && size_passed_back + node_data_space_size > free_size);
+
+        node_p = rec_set_p->get_prev (node_p);
+        node_data_space_p = rcs_get_node_data_space (rec_set_p, node_p);
+
+        free_rec_p = (record_t *) (node_data_space_p + node_data_space_size - \
+                                      (free_size - size_passed_back));
+      }
+
+      rcs_init_free_record (rec_set_p, free_rec_p, place_p, free_size);
+    }
+  }
+  else if (free_size != 0)
+  {
+    rcs_chunked_list_t::node_t *node_p = rec_set_p->get_node_from_pointer (place_p);
+    JERRY_ASSERT (node_p != NULL);
+
+    rcs_chunked_list_t::node_t *next_node_p = rec_set_p->get_next (node_p);
+
+    while (next_node_p != NULL)
+    {
+      node_p = next_node_p;
+      next_node_p = rec_set_p->get_next (node_p);
+    }
+
+    uint8_t *node_data_space_p = rcs_get_node_data_space (rec_set_p, node_p);
+    const size_t node_data_space_size = rcs_get_node_data_space_size ();
+
+    record_t *free_rec_p = (record_t *) (node_data_space_p + node_data_space_size - free_size);
+    rcs_init_free_record (rec_set_p, free_rec_p, place_p, free_size);
+  }
+} /* rcs_alloc_record_in_place */
+
+/**
+ * Allocate record of specified size.
+ *
+ * @return record identifier
+ */
+static record_t *
+rcs_alloc_space_for_record (record_set_t *rec_set_p, /**< recordset */
+                            record_t **out_prev_rec_p, /**< out: pointer to record, previous to the allocated,
+                                                        *   or NULL if the allocated record is the first */
+                            size_t bytes) /**< size */
+{
+  rcs_assert_state_is_correct (rec_set_p);
+
+  JERRY_ASSERT (JERRY_ALIGNUP (bytes, RCS_DYN_STORAGE_LENGTH_UNIT) == bytes);
+  JERRY_ASSERT (out_prev_rec_p != NULL);
```

Comment:
AFAIK: the comment on the argument says that it can be `NULL`, so which one is correct now?